torch/utils/launch.py

```diff
@@ -0,0 +1,805 @@
+
+class MultiInstanceLauncher(Launcher):
+    r"""
+     Launcher for single instance and multi-instance
+     """
+    def launch(self, args):
+        processes = []
+        cores = []
+        set_kmp_affinity = True
+        if args.core_list:  # user specify what cores will be used by params
+            cores = [int(x) for x in args.core_list.split(",")]
+            if args.ncore_per_instance == -1:
+                logger.error("please specify the '--ncore_per_instance' if you have pass the --core_list params")
+                exit(-1)
+            elif args.ninstances > 1 and args.ncore_per_instance * args.ninstances < len(cores):
+                logger.warning("only first {} cores will be used, but you specify {} cores in core_list".format(args.ncore_per_instance * args.ninstances, len(cores)))
+            else:
+                args.ninstances = len(cores) // args.ncore_per_instance
+
+        else:
+            if args.use_logical_core:
+                if args.node_id != -1:
+                    cores = self.cpuinfo.get_node_logical_cores(args.node_id)
+                else:
+                    cores = self.cpuinfo.get_all_logical_cores()
+                    # When using all cores on all nodes, including logical cores, setting KMP_AFFINITY disables logical cores. Thus, KMP_AFFINITY should not be set.
+                    set_kmp_affinity = False
+            else:
+                if args.node_id != -1:
+                    cores = self.cpuinfo.get_node_physical_cores(args.node_id)
+                else:
+                    cores = self.cpuinfo.get_all_physical_cores()
+            if not args.multi_instance and args.ninstances == -1 and args.ncore_per_instance == -1:
+                args.ninstances = 1
+                args.ncore_per_instance = len(cores)
+            elif args.multi_instance and args.ninstances == -1 and args.ncore_per_instance == -1:
+                args.throughput_mode = True
+            elif args.ncore_per_instance == -1 and args.ninstances != -1:
+                if args.ninstances > len(cores):
+                    logger.error("there are {} total cores but you specify {} ninstances; please make sure ninstances <= total_cores)".format(len(cores), args.ninstances))
+                    exit(-1)
+                else:
+                    args.ncore_per_instance = len(cores) // args.ninstances
+            elif args.ncore_per_instance != -1 and args.ninstances == -1:
+                args.ninstances = len(cores) // args.ncore_per_instance
+            else:
+                if args.ninstances * args.ncore_per_instance > len(cores):
+                    logger.error("Please make sure ninstances * ncore_per_instance <= total_cores")
+                    exit(-1)
+            if args.latency_mode:
+                logger.warning('--latency_mode is exclusive to --ninstances, --ncore_per_instance, --node_id and --use_logical_core. They won\'t take effect even they are set explicitly.')
```

Comment:
Please use double quotes consistently throughout the code
```suggestion
                logger.warning("--latency_mode is exclusive to --ninstances, --ncore_per_instance, --node_id and --use_logical_core. They won\'t take effect even they are set explicitly.")
```